### ISSUE TITLE
Starting work to improve accessibility in Jitsi Meet

### ISSUE THREAD

**Issue by Leimi** [jitsi/jitsi-meet#12379]
Hey Jitsi Meet team,

I work with [France Université Numérique](https://www.france-universite-numerique.fr/en/), a French public interest group making digital services for education. We and others use Jitsi Meet and we would like to contribute to improve its accessibility, on behalf of the French [Inter-departmental Directorate of Digital Affairs (french content)](https://www.numerique.gouv.fr/dinum/).

I know a lot of work was already done to make Jitsi Meet WCAG-compliant and it shows, the app is already greatly usable with screen readers and other assistive technologies :)

As of today, we audited the Jitsi Meet web client following WCAG rules (through the RGAA) and by testing with keyboard-only navigation and screen readers. We also collected feedback from regular users of Jitsi that use screen readers heavily.

<details>
<summary>What is this "RGAA"?</summary>
 
The RGAA (<span lang="fr">_Référentiel général d'amélioration de l'accessibilité_</span>) is the French government's set of web accessibility rules that you are legally required to follow when building or using a website or webapp as a public French organization. It is **exclusively** based on WCAG rules. The difference with WCAG is that the RGAA is an operational set of rules, that directly translates WCAG rules to HTML/CSS/JS specific guidelines.

</details>

**We noticed a few things that should be improved:**

- Some toggle buttons need to have more-specific labels for screen reader users. For example, the mute button is labelled _"Mute or unmute your microphone"_, which is quickly confusing when using a screen reader. Even if the button has a `aria-pressed` attribute, it doesn't help that much because the user doesn't know if pressing the button mutes, or unmutes, the microphone. Describing it as _"Mute your microphone"_ in one case or _"Unmute your microphone"_ in the other would greatly help screen reader users. This is one of the easiest change to make we saw, while also one of the most raised and blocking issue for people we talked to. Note that this was already discussed a bit in [this issue](https://github.com/jitsi/jitsi-meet/issues/3590).
Related PR: https://github.com/jitsi/jitsi-meet/pull/12426
- When _not_ using a screen reader, some buttons are not accessible via keyboard while that should be the case. You can notice this when trying to access via keyboard the buttons behind the _"More actions"_ menu.  
Related PR: https://github.com/jitsi/jitsi-meet/pull/12657
- Some buttons have missing labels, making screen reader users pretty confused about what they are trying to interact with. This is noticeable for example on the _"Performance settings"_ button
Related PR: https://github.com/jitsi/jitsi-meet/pull/12969
- The whole UI could benefit from using heading tags and/or `region` ARIA roles. They are often used by screen reader users to quickly jump between parts of the page. This would help navigate between the chat panel, the actions bar, the meeting participants panel, etc.  
Related PR: https://github.com/jitsi/jitsi-meet/pull/12427
- Some form inputs don't have actual labels technically tied to them, making it hard for screen reader users to use forms. This is visible in the _Profil_ and _Sound_ tabs of the _Settings_ dialog.  
Related PR: https://github.com/jitsi/jitsi-meet/pull/12969

- Information is sometimes hidden to screen reader users, like it doesn't exist. For example, in the meeting participants panel, the mute/camera indicators have no text alternative.
Related PR: https://github.com/jitsi/jitsi-meet/pull/12969
- Navigation with keyboard is sometimes confusing because the keyboard focus doesn't follow the navigation. For example, activating the _"Invite someone"_ button makes a dialog pop up, but the keyboard focus stays on the _"Invite someone"_ button. The focus should directly change to be in the dialog.  
Related PR: https://github.com/jitsi/jitsi-meet/pull/12657

This list is not exhaustive but a good representation of accessibility-related work that could be done in Jitsi Meet.
</details>

**We really want to contribute to fix all the issues we noticed and others, so that all users can fully use Jitsi Meet.**

Note that most issues we noticed would not require changes in the actual visible UI, or at least, not big changes.

We would like to start work on all these issues right now and proceed by submitting one pull request for each issue we listed above if that's ok with you.

**Please let us know what you think :)**

Also, any additional feedback on accessibity in Jitsi Meet is more than welcomed.

Thanks

**Comment by saghul** [comment]
Hi there @Leimi 

> Hey Jitsi Meet team,
> 
> I work with [France Université Numérique](https://www.france-universite-numerique.fr/en/), a French public interest group making digital services for education. We and others use Jitsi Meet and we would like to contribute to improve its accessibility, on behalf of the French [Inter-departmental Directorate of Digital Affairs (french content)](https://www.numerique.gouv.fr/dinum/).

This is great news, thanks for getting in touch!

We'll be happy to collaborate to get these resolved. Some responses below:



> * Some toggle buttons need to have more-specific labels for screen reader users. For example, the mute button is labelled _"Mute or unmute your microphone"_, which is quickly confusing when using a screen reader. Even if the button has a `aria-pressed` attribute, it doesn't help that much because the user doesn't know if pressing the button mutes, or unmutes, the microphone. Describing it as _"Mute your microphone"_ in one case or _"Unmute your microphone"_ in the other would greatly help screen reader users. This is one of the easiest change to make we saw, while also one of the most raised and blocking issue for people we talked to. Note that this was already discussed a bit in

Yep, this should be fixable y having different labels considering the current state of the button.

> * When _not_ using a screen reader, some buttons are not accessible via keyboard while that should be the case. You can notice this when trying to access via keyboard the buttons behind the _"More actions"_ menu.

You mean one can't navigate the entries in the more menu, after opening it? 

> * Some buttons have missing labels, making screen reader users pretty confused about what they are trying to interact with. This is noticeable for example on the _"Performance settings"_ button

👍 

> * The whole UI could benefit from using heading tags and/or `region` ARIA roles. They are often used by screen reader users to quickly jump between parts of the page. This would help navigate between the chat panel, the actions bar, the meeting participants panel, etc.

Today I learned :-) That sounds useful indeed!

> * Some form inputs don't have actual labels technically tied to them, making it hard for screen reader users to use forms. This is visible in the _Profil_ and _Sound_ tabs of the _Settings_ dialog.

I think we might be using fewer forms now, is there a way to link a label and an input together differently?

> * Information is sometimes hidden to screen reader users, like it doesn't exist. For example, in the meeting participants panel, the mute/camera indicators have no text alternative.

You mean the `alt` property? Yeah that makes sense.



> * Navigation with keyboard is sometimes confusing because the keyboard focus doesn't follow the navigation. For example, activating the _"Invite someone"_ button makes a dialog pop up, but the keyboard focus stays on the _"Invite someone"_ button. The focus should directly change to be in the dialog.

👍 



> This list is not exhaustive but a good representation of accessibility-related work that could be done in Jitsi Meet.

Thanks for putting the list together! It does show there is work to do. I didn't find any of the items in it particularly troublesome, they all sound like bugs we'd like to see fixed.



> We really want to contribute to fix all the issues we noticed and others, so that all users can fully use Jitsi Meet.

❤️ 



> Note that most issues we noticed would not require changes in the actual visible UI, or at least, not big changes.
> 
> We would like to start work on all these issues right now and proceed by submitting one pull request for each issue we listed above if that's ok with you.

Absolutely, please do! Making this change incrementally is the best way to go, so one PR per item sounds great. We can make faster progress on small PRs than on a gigantic one.

Thanks a lot for volunteering to work on this! 🚀

**Comment by Leimi** [comment]
Thanks for the warm welcome :)

> You mean one can't navigate the entries in the more menu, after opening it?

Via keyboard, yes.

When using a screen reader, the current situation is a bit annoying but not a real deal-breaker, since a screen reader makes any content on the page focusable, you manage to go through the menu. But when using a browser without any screen reader or other assistive technology, using only the "tab" key to focus interactive stuff, you can't access these entries.

> I think we might be using fewer forms now, is there a way to link a label and an input together differently?

Just to be clear, I said "forms" but it's something necessary to check on any form input (text, checkbox, radio, etc.), whether it's part of an actual `form` or not.

There are multiple ways to add a label to an input. You can have an `input` tag as a child of its `label` in the DOM, you can have `<label id="input-tag-id">`, you can set  `aria-label` or `aria-labelledby` attributes on `input` tags… I'll start working on this pretty soon actually :)

> You mean the `alt` property? Yeah that makes sense.

Well yes, but it's kind of like with form labels, there are multiple correct ways to add a "text alternative". Depending on the code context, sometimes one way is a better fit than the other. I didn't mention anything too specific on purpose but yeah most of the time, setting the `alt` is the straightforward way.

I'll start to open PRs soon :ok_hand:

**Comment by saghul** [comment]
> Via keyboard, yes.
> 
> When using a screen reader, the current situation is a bit annoying but not a real deal-breaker, since a screen reader makes any content on the page focusable, you manage to go through the menu. But when using a browser without any screen reader or other assistive technology, using only the "tab" key to focus interactive stuff, you can't access these entries.

I see. Yeah, that sounds like a bug that needs fixing.



> Just to be clear, I said "forms" but it's something necessary to check on any form input (text, checkbox, radio, etc.), whether it's part of an actual `form` or not.
> 
> There are multiple ways to add a label to an input. You can have an `input` tag as a child of its `label` in the DOM, you can have `<label id="input-tag-id">`, you can set `aria-label` or `aria-labelledby` attributes on `input` tags… I'll start working on this pretty soon actually :)

Got it. Thanks for working on this!



> Well yes, but it's kind of like with form labels, there are multiple correct ways to add a "text alternative". Depending on the code context, sometimes one way is a better fit than the other. I didn't mention anything too specific on purpose but yeah most of the time, setting the `alt` is the straightforward way.

👍 



> I'll start opening PRs soon 👌

Wonderful! Please @ me so I don't miss it and we can expedite reviews.

**Comment by damencho** [comment]
@Leimi is there an automated way for validation whether new code is correct and does not have these problems? This is something that will help for not breaking this after being fixed.

**Comment by damencho** [comment]
@saghul can we integrate the eslint one? Interesting is how many of the current mentioned issues are detected by that ...

**Comment by saghul** [comment]
Yeah, I like the eslint approach, we should give it a try!

**Comment by sampaccoud** [comment]
To add my 2 cents:
- maybe we could also mention accessibility guidelines in the "contributing" guide? @Leimi wrote one for [FUN](https://www.france-universite-numerique.fr/en) that could inspire you: https://handbook.openfun.fr/accessibility
- maybe adding comments to the code everywhere we fix issues would help make contributors aware and avoid breaking things too fast?

**Comment by saghul** [comment]
> To add my 2 cents:
> 
> - maybe we could also mention accessibility guidelines in the "contributing" guide? @Leimi wrote one for [FUN](https://www.france-universite-numerique.fr/en) that could inspire you: https://handbook.openfun.fr/accessibility
> 
> - maybe adding comments to the code everywhere we fix issues would help make contributors aware and avoid breaking things too fast?

Makes sense!

Our contributing guidelines are here: https://jitsi.github.io/handbook/docs/dev-guide/dev-guide-contributing

We take PRs on the handbook, so a section about accessibility would indeed be welcome.

**Comment by keda82** [comment]
We have done quite a comprehensive WCAG review of the Jitsi UI with reference to WCAG 2.1 AA and EN 301 549 for each issue we have found. It also covers suggestions on how to adress certain issues.

our main issue now is that we do not have any developers on stand by ready to dig in to this. If someone in the community is already looking at this maybe our review is of interest

**Comment by saghul** [comment]
Most certainly! Is this something you can share?

**Comment by Leimi** [comment]
Hey, great :) I'm actually already working on this with open PRs (@saghul don't hesitate to check :P) and will continue so I guess that's pretty convenient haha. I already done some kind of audit based on wcag too. Sharing your work could indeed help 👌

**Comment by keda82** [comment]
Great job @Leimi ! Should be able to share, let me see how and get back asap 👍🏻😊

**Comment by keda82** [comment]
Is there a tag for accessibility we can use? Planning to post the issues we identified here on Github. Would be great if we can keep them ordered and avoid duplicates.

**Comment by saghul** [comment]
We actually have an "accessibility" label.

I just tagged this issue with it.

**Comment by keda82** [comment]
ah, great, missed that

**Comment by saghul** [comment]
Back to the original topic, @Leimi do you have plans to resume the work? Is anything on our side blocking you?

**Comment by Leimi** [comment]
Hey, yes sorry for giving no news, I definitely have plans to resume working on this very soon, the beginning of the year was really busy on my side… Actually wanted to continue this week if I could, or in february in any case.

**Comment by kenneth-edwall** [comment]
our WCAG review is ready to be published to the community. The plan is to report all the identified issues(50+ identified) and then start working on resolving them one-by-one. Instead of keeping it closed for our selves we figured it is better to share with the community to avoid "dubble work" and to get more people to dig in. 

So, how should be do this in the best way?
- Report each separate identified WCAG issue as a separate issue with accessability tag or all issues in on github issue. Guess the first option?
- How would our developers contribute in the best and organised way, pull requests to master? fork?
- It is also critical that we establish a long term solution to avoid WCAG improvements being broken in the future. How do we move towards this goal?

This review is based on the EU web accessability directive, https://digital-strategy.ec.europa.eu/en/policies/web-accessibility, and is put high on the agenda for many organisations.

**Comment by saghul** [comment]
Do you work with @Leimi or is this a parallel effort?

**Comment by kenneth-edwall** [comment]
Its most likely both, could be duplicates to what @Leimi is working on.

@Leimi , maybe you and I should sync before we start posting our issues? I sent you an email to manu@habite.la

**Comment by saghul** [comment]
That would be very useful, yeah.

**Comment by Leimi** [comment]
Hey, did talk with @kenneth-edwall via email, great to see more people helping with this :) Will myself continue work on my open PRs (and more) starting at the end of the week and normally through at least the end of february.

**Comment by saghul** [comment]
Excellent, thanks a lot for the update!

**Comment by Leimi** [comment]
Hey, sorry for the huge delay…

This PR is almost finished on my part and should fix lots of remaining issues:

https://github.com/jitsi/jitsi-meet/pull/12969

After that I have a couple things in mind to check, and from my point of view most of the important accessibility-related issues should be resolved with all this.

**Comment by Leimi** [comment]
So, update time!

There are a few stuff I need to check after #12969:

* [ ] there are still a couple places in the app where I think content is not accessible via screen reader or keyboard nav. Mostly in the participants pane. Also I'm quite sure some things I didn't even get to test because there are quite a 
* [ ] dialogs seem to have issues for screen reader users. For people just using keyboards, it works okay: pressing tab will focus elements in the dialog, and loop in the dialog until pressing escape. For people using screen readers and maybe navigating with arrow keys or other shorcuts, focus will not be forced to stay in the dialog. This can be problematic. Another issue is that sometimes, on dialog appearance, focus is not automatically set to/inside the dialog. Pressing tab _will_ set focus into it, but it doesn't do anything on appearance. Sometimes. This combined with the issue that screen reader users can navigate outside, makes it really easy to just don't even notice a dialog just opened. And in that case it's a big problem.
* [ ] we need to improve tooltips with keyboard navigation. Meaning, when focusing an icon in the toolbox, the tooltip needs to appear. Like when hovering an icon with the mouse. I tried to look how to implement that a few weeks ago but remember it seemed more difficult than I thought.
* [ ] there are a few things to change in the index.html code in order to improve a11y. For example, adding a doctype or lang attribute. But I couldn't figure out how to test this locally (it seems the default dev env doesn't get the local index.html?). any tips @saghul? :innocent: 

That's the main remaining big things to do from my point of view, and I feel a few of them are a bit tricky. I'll see.

There are still plenty things we could improve after that, for example: there are a few places where "radio" inputs are implemented by hand, with ARIA roles and attributes. This works okay, but keyboard nav is not ideal with those for now. If we simulate radios and tell the browser (and screen readers) we do, we should handle keyboard nav accordingly, meaning we can go back and forth radio items with arrow keys, for example. This is not done at all.  
This is a typical example of something that doesn't follow best practices and should be resolved in an ideal world. But from my point of view they are not deal breakers (but my point of view is debatable for sure). While it is more cumbersome, it doesn't prevent users from accessing things. So I don't think I will spend time on this issue and others like it.
